android/src/main/java/io/grpc/android/UDSSocket.java

```diff
@@ -0,0 +1,304 @@
+/*
+ * Copyright 2021 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.android;
+
+import android.net.LocalSocket;
+import android.net.LocalSocketAddress;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import java.io.FilterInputStream;
+import java.io.FilterOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.InetAddress;
+import java.net.Socket;
+import java.net.SocketAddress;
+import java.net.SocketException;
+import java.nio.channels.SocketChannel;
+
+@SuppressWarnings("UnsynchronizedOverridesSynchronized") // Rely on LocalSocket's synchronization
+class UDSSocket extends Socket {
+
+  private final LocalSocket localSocket;
+  private final LocalSocketAddress localSocketAddress;
+
+  @GuardedBy("this")
```

Comment:
Should be GuardedBy `localSocket`, otherwise calling `getSendBufferSize()` and `close()` concurrently would deadlock.

Reply:
nvm, it won't deadlock because localSocket.getSendBufferSize() is just calling localSocket.isClosed(), not UDSSocket.isClosed().